ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs

```diff
@@ -74,23 +84,33 @@ type IProxyTypeProvidersManager =
         systemRuntimeAssemblyVersion: Version *
         compilerToolsPath: string list -> ITypeProvider list
 
+    abstract member HasGenerativeTypeProviders: project: IProject -> bool
     abstract member Dump: unit -> string
 
-type TypeProvidersManager(connection: TypeProvidersConnection) =
+type TypeProvidersManager(connection: TypeProvidersConnection, fcsProjectProvider: IFcsProjectProvider) =
     let protocol = connection.ProtocolModel.RdTypeProviderProcessModel
     let lifetime = connection.Lifetime
     let tpContext = TypeProvidersContext(connection)
     let typeProviders = TypeProvidersCache()
+    let lock = SpinWaitLock()
+    let doesProjectContainGenerativeProviders = Dictionary<IProject, bool>()
+
+    do
+        connection.Execute(fun () ->
+            protocol.Invalidate.Advise(lifetime, fun id -> typeProviders.Get(id).OnInvalidate()))
 
-    do connection.Execute(fun () ->
-        protocol.Invalidate.Advise(lifetime, fun id -> typeProviders.Get(id).OnInvalidate()))
+        fcsProjectProvider.ModuleInvalidated.Advise(lifetime, fun psiModule ->
+            let project = getModuleProject psiModule
+            try lock.Enter()
+                doesProjectContainGenerativeProviders.Remove(project.NotNull()) |> ignore
+            finally lock.Exit())
```

Comment:
Could this and other `lock` usages be simpler with `JetFastSemiReenterableRWLock`?